pkg/monitoring/chain_reader.go

```diff
@@ -34,14 +34,54 @@ type chainReader struct {
 	contractStoreSequencer sync.Mutex
 }
 
-func (c *chainReader) TxsEvents(events []string, paginationParams *query.PageRequest) (*txtypes.GetTxsEventResponse, error) {
+func (c *chainReader) TxsEvents(ctx context.Context, events []string, paginationParams *query.PageRequest) (*txtypes.GetTxsEventResponse, error) {
 	c.txEventsSequencer.Lock()
 	defer c.txEventsSequencer.Unlock()
-	return c.client.TxsEvents(events, paginationParams)
+	raw, err := withContext(ctx, func() (interface{}, error) {
+		return c.client.TxsEvents(events, paginationParams)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return raw.(*txtypes.GetTxsEventResponse), nil
 }
 
-func (c *chainReader) ContractStore(contractAddress sdk.AccAddress, queryMsg []byte) ([]byte, error) {
+func (c *chainReader) ContractStore(ctx context.Context, contractAddress sdk.AccAddress, queryMsg []byte) ([]byte, error) {
 	c.contractStoreSequencer.Lock()
 	defer c.contractStoreSequencer.Unlock()
-	return c.client.ContractStore(contractAddress, queryMsg)
+	raw, err := withContext(ctx, func() (interface{}, error) {
+		return c.client.ContractStore(contractAddress, queryMsg)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return raw.([]byte), nil
+}
+
+type callResult struct {
+	data interface{}
+	err  error
+}
+
+// Helpers
+
+// withContext makes a function that does not take in a context exit when the context cancels or expires.
+// In reality withContext will abandon a call that continues after the context expires and simply return an error.
+// This helper is needed, because the cosmos/tendermint clients don't respect context.
+// Note! This method may leak goroutines.
```

Comment:
Do you have an open issue/link for this?

Reply:
Nope. Will make one